### DISCUSSION TITLE
[Xlsx-Renderer] Config builder

### DISCUSSION THREAD

**Discussion by Siemienik** [Siemienik/XToolset#173] [Ideas]
# The Config Builder - still DRAFT

_Please do not start until it is a draft_

## Place of live

`./src/configBuilder/`

## Sample usages
```js

const cfg = ListCfgBuilder.createList()
  .worksheet('asd')
  .column('key', 1 ,integerMapper)
  .column('key2' 2, splitMapper.separator(':'))
  .toConfig(); // or .build() - or event better, put this cfg directly into getAllItems() 

// and
const cfg2 =ListCfgBuilder.from(cfg)
  .worksheet('second')
  .getConfig();


// json support
const asJson = CfgBuilder.from(cfg).toJson();

const fromJson = CfgBuilder.from(asJson, {type: 'json'}).toConfig();
```

## TODO

* [ ] ListVerticalConfigBuilder
* [ ] SingleObjectConfigBuilder
* [ ] Research required: ConfigBuilder for both
* [ ] toJson/fromJson (ConfigWriter/ConfigReader)
    * [ ] string representation for mappers

## Draft's todo:

* [ ] analysis discriminator for reading config
* [ ] become sure about APIs
* [ ] create issues for each TODO Point
* [ ] consider about moving ConfigWriter & Reader out of scope of this issue
* [ ] write about motivations to add it into project

**Comment by Metastasis** [comment]
@Siemienik 
Could you add a little description?

**Comment by Siemienik** [comment]
EDIT: I copied this comment into issue description

<details>

# The Config Builder - still DRAFT

_Please do not start until it is a draft_

## Place of live

`./src/configBuilder/`

## Sample usages
```js

const cfg = ListCfgBuilder.createList()
  .worksheet('asd')
  .column('key', 1 ,integerMapper)
  .column('key2' 2, splitMapper.separator(':'))
  .toConfig(); // or .build() - or event better, put this cfg directly into getAllItems() 

// and
const cfg2 =ListCfgBuilder.from(cfg)
  .worksheet('second')
  .getConfig();


// json support
const asJson = CfgBuilder.from(cfg).toJson();

const fromJson = CfgBuilder.from(asJson, {type: 'json'}).toConfig();
```

## TODO

* [ ] ListVerticalConfigBuilder
* [ ] SingleObjectConfigBuilder
* [ ] Research required: ConfigBuilder for both
* [ ] toJson/fromJson (ConfigWriter/ConfigReader)
    * [ ] string representation for mappers


## Draft's todo:

* [ ] analysis discriminator for reading config
* [ ] become sure about APIs
* [ ] create issues for each TODO Point
* [ ] consider about moving ConfigWriter & Reader out of scope of this issue
* [ ] write about motivations to add it into project

</details>

**Comment by Siemienik** [comment]
I've done research, however I wait for #62 and #63 to be 100% sure that everything is included.
